### PR TITLE
ENH: Add `replace` method to `Index` 

### LINKED ISSUE #19495: [Feature Request] Add `replace` method to `Index` objects
I usually have to change the value of an element in `pd.Index` but it's difficult without the `replace` method that is present in `pd.Series`.  I usually have a bunch of `Jupyter` cells depending on a particular `pd.DataFrame` or `pd.Series` with complicated transformations and a graph at some point.  In most cases, I have to change the labels for a few to clean them up for publication quality figures.  Currently, I have to use an if else statement with `map` method but it would nice to be able to just do `replace`. 

```python
print(pd.__version__)
print("_________________________")
idx = pd.Index([0,1,2])
Se = pd.Series(list("abc"), index=idx)
# Can do it to pd.Series
print(Se)
print(Se.replace({"b":"replaced_b"}))
# but not pd.Index
print("_________________________")
print(idx)
d = {1:100}
print(idx.map(lambda x: d[x] if x in d else x))
print(idx.replace({1:100}))

# 0.21.1
# _________________________
# 0    a
# 1    b
# 2    c
# dtype: object
# 0             a
# 1    replaced_b
# 2             c
# dtype: object
# _________________________
# Int64Index([0, 1, 2], dtype='int64')
# Int64Index([0, 100, 2], dtype='int64')
# ---------------------------------------------------------------------------
# AttributeError                            Traceback (most recent call last)
# <ipython-input-383-0b8eacd67c54> in <module>()
#      11 d = {1:100}
#      12 print(idx.map(lambda x: d[x] if x in d else x))
# ---> 13 print(idx.replace({1:100}))
#      14 

# AttributeError: 'Int64Index' object has no attribute 'replace'
```

### PR DESCRIPTION
**Pull request by oguzhanogreden**
- [X] closes #19495
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] docstring for the new Index.replace() method
- [X] won't do: type hints
- [ ] merge with `ENH:` commit message

Added a replace method to Index classes, as well as tests.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
 
+    def replace(
+        self,
+        to_replace=None,
+        value=None,
+        inplace=False,
+        limit=None,
+        regex=False,
+        method="pad",
+    ):
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Can you add a docstring

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
         """
         return self.set_names([name], inplace=inplace)
 
+    def replace(
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Can you add type annotations

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
+                    "This is implemented in NDFrame.replace(). However,"
+                    "not clear if we should include this in the API."
+                    "See issue 5319 and PR 5600. But also note that this"
+                    "use is not mentioned in the docs."
+                )
+
+            if not is_dict_like(to_replace):
+                if not is_dict_like(regex):
+                    raise TypeError(
+                        'If "to_replace" and "value" are both None '
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Can you invert the quotes, meaning that ```to_replace``` will be surrounded by single quotes.

(Just to be consistent with the ```AssertionError```'s error message above)

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
+                to_replace=to_replace,
+                value=value,
+                inplace=inplace,
+                limit=limit,
+                regex=regex,
+            )
+
+        else:
+            if is_dict_like(to_replace):
+                raise TypeError("If `to_replace` is a dict, `value` should be None.")
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Replace backticks with single quotes.

(To he consistent with the error messages)

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
+                        method=method,
+                    )
+            else:
+                new_index = self._replace_single(
+                    to_replace=to_replace, value=value, regex=regex,
+                )
+
+        return new_index
+
+    def replace_list(self, src_list, dest_list, regex=False):
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Type annotations if possible

File: pandas/core/indexes/base.py [hunk 6]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
+            If true, perform regular expression substitution.
+
+        Returns
+        -------
+        Index
+            The same type as the caller.
+        """
+
+        from pandas.core.internals.managers import _compare_or_regex_search
+        from pandas.core.internals.managers import maybe_convert_objects
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Unify this two imports

File: pandas/core/indexes/base.py [hunk 7]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
+        zipped = zip(src_list, dest_list)
+        for i, (_, dest) in enumerate(zipped):
+            m = masks[i]
+            if m.any():
+                new_index = new_index.putmask(mask=m, value=dest)
+
+        return new_index
+
+    def _replace_single(
+        self, to_replace, value, regex=False, convert=True, mask=None,
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Type annotations

File: pandas/tests/indexes/base_class/test_replace.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    expected = pd.Index(["a", 2, "c"])
+
+    result = index.replace({1: "a", 3: "c"})
+    tm.assert_equal(result, expected)
+
+
+def test_index_replace_2_1():
+    index = pd.Index([1, 2, 3])
+
+    with pytest.raises(TypeError):
```
Comment (MomIsBestFriend) [inline pandas/tests/indexes/base_class/test_replace.py]: Can you can the ```match``` argument, and the expected error message

File: pandas/tests/indexes/base_class/test_replace.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        index.replace({1: "a", 3: "c"}, "x")
+
+
+def test_index_replace_3():
+    index = pd.Index([1, None, 2])
+    with pytest.raises(NotImplementedError):
```
Comment (MomIsBestFriend) [inline pandas/tests/indexes/base_class/test_replace.py]: Can you add the ```match``` argument, and the expected error message

File: pandas/core/indexes/base.py [hunk 10]
```diff
@@ -1339,6 +1342,237 @@ def rename(self, name, inplace=False):
...
+        else:
+            if is_dict_like(to_replace):
+                raise TypeError("If `to_replace` is a dict, `value` should be None.")
+            if is_list_like(to_replace):
+                if is_list_like(value):
+                    if len(to_replace) != len(value):
+                        # NOTE: Corresponding error message in core.generic.replace
+                        #       is not clear. Let's decide on one.
+                        raise ValueError(
+                            f"Length of `to_replace=` ({len(to_replace)}) should "
```
Comment (MomIsBestFriend) [inline pandas/core/indexes/base.py]: Replace backticks with single quotes

(To be consistent with the error messages)

**Review by ShaharNaveh — CHANGES_REQUESTED** [review]
@oguzhanogreden Thank you so much for taking this!

Overall this looks really good!


---

Some minor comments :)

**Comment by MomIsBestFriend** [comment]
Also a question, can any of the new added tests be parameteraize?

File: pandas/tests/indexes/base_class/test_replace.py (hunk 9 above)
Comment (oguzhanogreden) [reply pandas/tests/indexes/base_class/test_replace.py]: Are we sure we want to raise a `NotImplementedError` here? I'd rather not receive one from a pandas. I'll wait until that's resolved and add the message accordingly.

**Comment by oguzhanogreden** [comment]
Thanks @MomIsBestFriend, helpful indeed. I'll be waiting for a green flag from the core before I turn to docstrings and type annotations. Mainly since they take a lot of time... 😇  namely (1) figuring out the mechanics of docstrings for different subclasses and (2) correct types for type annotations.

**Comment by MomIsBestFriend** [comment]
@oguzhanogreden ATM you can ignore the CI failure of ```CI/Web And Docs```, we haven't yet resolved this issue. so you can ignore it. (for now at least)

Just be sure to merge master every now and then :)

**Review by datapythonista — COMMENTED** [review]
Did you check if your implementation is faster than:
```python
import pandas

def replace(self, *args, **kwargs):
    return pandas.Index(self.to_series().replace(*args, **kwargs))
pandas.Index.replace = replace

idx = pandas.Index([1, 2, 3, 4, 5, 6])
idx.replace(3, 9)
```

Are there cases when converting to `Series` is a problem?

Seems like a huge complexity added that we can hopefully avoid.

**Comment by oguzhanogreden** [comment]
Regarding the speed point, this seems faster. At the very least, faster more often. Be warned I'm not on top of what's going on `%timeit` mechanics. I chose the arguments below since they seemed reasonable from a measurement perspective. If there's a gotcha about performance, do let me know:

``` python
# Running on a MacBook Pro from 2017, under some use load
# (browsers with many tabs are open etc.)

import numpy as np

import pandas as pd

idx = pd.Index([1, 2, 3, 4, 5, 6])
%timeit -n 10000 -r 50 idx.replace(3, 9)

# 69.9 µs ± 6.67 µs per loop (mean ± std. dev. of 50 runs, 10000 loops each)

def replace(self, *args, **kwargs):
    return pd.Index(self.to_series().replace(*args, **kwargs))
pd.Index.replace = replace

idx = pd.Index([1, 2, 3, 4, 5, 6])
%timeit -n 10000 -r 50 idx.replace(3, 9)

# 348 µs ± 102 µs per loop (mean ± std. dev. of 50 runs, 10000 loops each)
# ^^^ couldn't get higher precision here, also not a much lower point estimate
```

Happy to move this to `asv` if the output is considerably more reliable.

**Comment by oguzhanogreden** [comment]
Regarding the complexity point - agreed. Though I think there's some benefit to having these methods available. I found myself in a situation where I "just wanted to be able to do what I know from other parts of the pandas API". It's up to you folks to judge that against maintenance burden, ultimately.

I'd be interested in a contributing to a refactor around these methods. However I can imagine it'll be pretty tricky to abstract things away. Considering the nitty-gritty involved...

I'll survey properly whether i `.to_series()` fails to solve anything in the coming days.

Finally, I just added a few `NotImplementedError`s where this method, as it stands, shouldn't return anything. Eventually I'm happy to address those, at a cost of some more complexity, if the overall judgement is these are useful to have upstream.

**Comment by datapythonista** [comment]
Just to be clear, my proposal was not that you use that code (with the `to_series()`) in your own code. But that the implementation in this PR use it instead of all the code you're adding.

So, you'll still get the same functionality in pandas `Index.replace()`, but instead of adding 300 lines of code here, you'll be adding 30. But seems like the difference in performance is quite big.

**Comment by oguzhanogreden** [comment]
Thanks. That was clear. I made the point about complexity since it's not clear to me how to weigh the performance difference against growing the code base so much :)

I suggest we follow up once I do the following:

- [ ] check if `to_series()` can handle most of the requirements,
- [ ] try and identify the cause of the performance issue, see if I can address and wrap `.to_series()` to create a solution,

**Comment by jbrockmendel** [comment]
I think we'd be much better off with

```
def replace(self, whatever):
    result = self.to_series().replace(whatever)
    return Index(result)
```

**Comment by mroeschke** [comment]
@oguzhanogreden do you have time to merge in master and adapt the implementation @jbrockmendel proposed? Would be beneficial to have a shared implementation to reduce code maintenance.

**Comment by oguzhanogreden** [comment]
Sorry, I didn't get to prioritize this for a while. I'll try to do that in a few days, with the suggested solution.

File: pandas/core/indexes/base.py [hunk 11]
```diff
@@ -871,7 +871,7 @@ def _format_data(self, name=None) -> str_t:
         if self.inferred_type == "string":
             is_justify = False
         elif self.inferred_type == "categorical":
-            if is_object_dtype(self.categories):  # type: ignore
```
Comment (oguzhanogreden) [inline pandas/core/indexes/base.py:874]: FYI, I had to remove this due to mypy complaints.

**Comment by oguzhanogreden** [comment]
I added some tests for the "basic functionality". Not quite sure how to strike the balance: one arguments says all the existing tests of Index/MultiIndex -> Series/DataFrame conversions and the relevant `replace()` methods ensures the functionality here. What do you think?

I'll move on to type hints and try to figure out how to reuse, if possible at all, the docstrings.

File: pandas/core/indexes/base.py (hunk 11 above)
Comment (oguzhanogreden) [reply pandas/core/indexes/base.py:874]: Tests failed due to this line. However, `mypy` run as part of git-commit complains here saying "unused 'type: ignore' comment".

**Comment by oguzhanogreden** [comment]
I'll consider type hints out of scope for this PR since it will require a refactor. For instance, mypy will complain for `to_replace.items()` since `to_replace=` can be both a `list` and a `dict`.

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (oguzhanogreden) [reply pandas/core/indexes/base.py:1356]: https://github.com/pandas-dev/pandas/pull/32542#issuecomment-646993103

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (oguzhanogreden) [reply pandas/core/indexes/base.py:1354]: Won't do: https://github.com/pandas-dev/pandas/pull/32542#issuecomment-646993103

File: pandas/core/indexes/base.py (hunk 5 above)
Comment (oguzhanogreden) [reply pandas/core/indexes/base.py]: Won't do: https://github.com/pandas-dev/pandas/pull/32542#issuecomment-646993103

**Comment by jreback** [comment]
if u can merge master and get passing will have a look

**Comment by oguzhanogreden** [comment]
Hopefully the shared docs make sense.

Please be aware that the documentation of `replace()` methods contains has examples for DataFrame and Series only. I did _not_ add any more examples since the examples are fairly extensive already.

**Comment by oguzhanogreden** [comment]
@jreback, all green.

File: pandas/core/common.py [hunk 12]
```diff
@@ -34,6 +34,306 @@
 from pandas.core.dtypes.inference import _iterable_not_string
 from pandas.core.dtypes.missing import isna, isnull, notnull  # noqa
 
+_core_shared_docs: Dict[str, str] = dict(
```
Comment (jbrockmendel) [inline pandas/core/common.py:37]: could go in core.shared_docs?

Comment (oguzhanogreden) [reply pandas/core/common.py]: Moved and merged master again.

**Comment by oguzhanogreden** [comment]
The failing tests are due to master, AFAICT. If this isn't merged until the master is green, I will re-merge it once I notice the master passing.

(Just to have this noted down: The last commit message should read "~Cache~Catch errors not caught by checks")

**Comment by oguzhanogreden** [comment]
@jreback looking at the latest merges, I have impression that the remaining failing test is 'okay'? 😬

**Comment by jreback** [comment]
@oguzhanogreden I haven't looked at your PR yet at all, so no idea.

File: doc/source/whatsnew/v1.1.0.rst [hunk 13]
```diff
@@ -341,6 +341,7 @@ Other enhancements
 - :meth:`DataFrame.agg` and :meth:`Series.agg` now accept named aggregation for renaming the output columns/indexes. (:issue:`26513`)
 - ``compute.use_numba`` now exists as a configuration option that utilizes the numba engine when available (:issue:`33966`, :issue:`35374`)
 - :meth:`Series.plot` now supports asymmetric error bars. Previously, if :meth:`Series.plot` received a "2xN" array with error values for `yerr` and/or `xerr`, the left/lower values (first row) were mirrored, while the right/upper values (second row) were ignored. Now, the first row represents the left/lower error values and the second row the right/upper error values. (:issue:`9536`)
+- :class:`Index` and :class:`MultiIndex` now has a `replace()` method (:issue:`19495`).
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:344]: move to 1.2

**Review by jreback — CHANGES_REQUESTED** [review]
just a tiny comment so far; there is a lot going on here and will need review

**Comment by dsaxton** [comment]
@oguzhanogreden Can you merge master and fix merge conflicts?

File: pandas/core/indexes/category.py [hunk 14]
```diff
@@ -769,5 +769,8 @@ def _wrap_joined_index(
         name = get_op_result_name(self, other)
         return self._create_from_codes(joined, name=name)
 
+    def replace(self, *args, **kwargs):
```
Comment (jbrockmendel) [inline pandas/core/indexes/category.py:772]: can you make the signature match Index.replace anyway

possible follow-up: Categorical _does_ implement replace, so it should be easy to pass-through here

**Comment by jbrockmendel** [comment]
@oguzhanogreden can you rebase and address comments; i think this was fairly close to ready

**Comment by oguzhanogreden** [comment]
#36178 happened but `_shared_doc` is still used across the code base. I'll shortly open a PR for that.

File: pandas/core/indexes/category.py (hunk 14 above)
Comment (oguzhanogreden) [reply pandas/core/indexes/category.py]: Signature matches now.

The implementation of pd.Categorical isn't able to support the replace operation, since it defers part of the replace logic [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/categorical.py#L2249-L2250). I tried the `pd.CategoricalIndex().to_series()` route but that failed due to #37899.

I've seen a few more issues with Categorical related replaces, therefore suggest pushing this over the line and deferring `CategoricalIndex.replace()` to after those are fixed. If agreed, I'll create an issue for noting the lack of implementation here.

File: doc/source/whatsnew/v1.1.0.rst (hunk 13 above)
Comment (oguzhanogreden) [reply doc/source/whatsnew/v1.1.0.rst]: Done.

File: pandas/core/indexes/category.py [hunk 15]
```diff
@@ -697,3 +697,15 @@ def _delegate_method(self, name: str, *args, **kwargs):
...
+        self,
+        to_replace=None,
+        value=None,
+        limit=None,
+        regex=False,
+        method="pad",
+    ):
+        raise NotImplementedError(
+            "Replacing values of a CategoricalIndex is not supported."
+        )
```
Comment (jbrockmendel) [inline pandas/core/indexes/category.py:711]: remind me why we need to disable this?

File: pandas/core/shared_docs.py [hunk 16]
```diff
@@ -145,115 +145,115 @@
...
-0  a        B      1
-1  b        B      3
-2  c        B      5
-
->>> %(caller)sid_vars=[('A', 'D')], value_vars=[('B', 'E')])
-  (A, D) variable_0 variable_1  value
-0      a          B          E      1
-1      b          B          E      3
-2      c          B          E      5
+    Unpivot a DataFrame from wide to long format, optionally leaving identifiers set.
```
Comment (jbrockmendel) [inline pandas/core/shared_docs.py:148]: this is unrelated?

Comment (oguzhanogreden) [reply pandas/core/shared_docs.py:148]: Ah, that's embarassing - reverting it.

File: pandas/core/shared_docs.py [hunk 17]
```diff
@@ -325,3 +325,304 @@
...
+    1   new  new
+    2  bait  xyz
+
+    Note that when replacing multiple ``bool`` or ``datetime64`` objects,
+    the data types in the `to_replace` parameter must match the data
+    type of the value being replaced:
+
+    >>> df = pd.DataFrame({{'A': [True, False, True],
+    ...                    'B': [False, True, False]}})
+    >>> df.replace({{'a string': 'new value', True: False}})  # raises
```
Comment (oguzhanogreden) [inline pandas/core/shared_docs.py:578]: This is one of the failing doctests.

The behavior seems to have changed in the `master` as well, I'll change the docstring and add a test case to fix behavior prospectively.

File: pandas/core/indexes/category.py (hunk 15 above)
Comment (oguzhanogreden) [reply pandas/core/indexes/category.py:666]: Otherwise it hits #37899.

Comment (oguzhanogreden) [reply pandas/core/indexes/category.py:670]: What do you think?

**Comment by oguzhanogreden** [comment]
(I can't replicate the build errors locally.)

File: pandas/core/indexes/category.py (hunk 15 above)
Comment (jbrockmendel) [reply pandas/core/indexes/category.py:670]: So some cases will hit the same bug that affects Series.replace and other cases will work fine (assuming we don't raise here)?  If so, then we're better off matching Series behavior and allowing the subset of cases that do work.

Comment (oguzhanogreden) [reply pandas/core/indexes/category.py:674]: I disabled `regex=` due to #38447 . Rest is enabled and added tests are added.

File: pandas/core/indexes/category.py [hunk 18]
```diff
@@ -655,3 +655,24 @@ def _delegate_method(self, name: str, *args, **kwargs):
         if is_scalar(res):
             return res
         return CategoricalIndex(res, name=self.name)
+
+    def replace(
```
Comment (oguzhanogreden) [inline pandas/core/indexes/category.py:659]: Not adding @doc here since it will require a lot of conditionals due to regex being disabled. We can add it after `replace()` is sorted out.

**Comment by jreback** [comment]
@oguzhanogreden can you do a pre-cursor PR that moves things and doesn't change anything (except for imports and so on), no functionailty at all. Its almost impossible to tell what you are changing otherwise.

**Comment by oguzhanogreden** [comment]
Now that #38561 is done, what's going here is a bit clearer. Failing test is also failing on master, not related to changes here.

Let me try to summarise to help with review:

1. The key change is the addition of `.replace()` method to Index classes.
2. pd.Categorical does not allow regex replaces (see https://github.com/pandas-dev/pandas/issues/38447), therefore [we're disabling](https://github.com/pandas-dev/pandas/pull/32542/files#diff-d510328d75f258e42e90cab852f2cd17cc85c7ee10e6c3f69734e1e5bc87801eR640) it.
3. Due to the same reason, I have [**not** added](https://github.com/pandas-dev/pandas/pull/32542/files#r541932202) documentation to CategoricalIndex. This helps keep shared docs simple (i.e. without a conditional for regex stuff). Once the issues with categorical are solved, we can just add the decorator and it's good to go. I'll make an issue and reference it where needed, once this is merged.

Thanks for your patience here! It took too long in part due to my wrong initial approach (huge PR) and in part due to my absence from last March. Looking forward to getting this out of the door and paying back the investment ;)

**Comment by jreback** [comment]
ok can you merge master and will look

**Comment by mroeschke** [comment]
This was looking pretty close but looks like this PR has gotten stale. Going to close but if interested in continuing please ping, merge master, and target this PR for the next release (1.3 as of writing this comment).
